storage/provider/openstack.go

```diff
@@ -0,0 +1,399 @@
+
+	// In most cases, it is quicker to get all volumes and loop
+	// locally than to make several round-trips to the provider.
+	cinderVols, err := s.openstackClient.GetVolumesSimple()
+	if err != nil {
+		return nil, err
+	}
+
+	// Map volume ID to volume info for faster lookup.
+	idToCinderVol := make(map[string]storage.Volume)
+	for _, volInfo := range cinderVols {
+		idToCinderVol[volInfo.VolumeId] = volInfo
+	}
+
+	volumes := make([]storage.Volume, len(volIds))
+	for volIdx, volId := range volIds {
+		cinderVol, ok := idToCinderVol[volId]
+		if !ok {
+			return nil, errors.Errorf("could not find volume: %s", volId)
+		}
+		volumes[volIdx] = cinderVol
+	}
+
+	return volumes, nil
+}
+
+// DestroyVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) DestroyVolumes(volIds []string) []error {
+
+	errors := make([]error, len(volIds))
+	for volIdx, volId := range volIds {
+		if err := s.openstackClient.DeleteVolume(volId); err != nil {
+			errors[volIdx] = err
+		}
+	}
+	return errors
+}
+
+// ValidateVolumeParams implements storage.VolumeSource.
+func (s *openstackVolumeSource) ValidateVolumeParams(params storage.VolumeParams) error {
+	return nil
+}
+
+// AttachVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) AttachVolumes(args []storage.VolumeAttachmentParams) ([]storage.VolumeAttachment, error) {
+
+	volAttachments := make([]storage.VolumeAttachment, len(args))
+	for argIdx, attachArg := range args {
+		if attachArg.Provider != CinderProviderType {
+			logger.Debugf(
+				"openstack storage provider was passed a request for another provider type: %v",
+				attachArg.Provider,
+			)
+			continue
+		}
+
+		// Check to see if the volume is already attached.
+		volAttachments, err := s.openstackClient.ListVolumeAttachments(string(attachArg.InstanceId))
+		if err != nil {
+			return nil, err
+		}
+		if findAttachment(attachArg.VolumeId, volAttachments) != nil {
+			continue
+		}
+
+		if err := <-s.openstackClient.VolumeStatusNotifier(
+			attachArg.VolumeId,
+			"available",
+			10,
+			5*time.Second,
+		); err != nil {
+			return nil, err
+		}
+
+		attachment, err := s.openstackClient.AttachVolume(
+			string(attachArg.InstanceId),
+			attachArg.VolumeId,
+			autoAssignedMountPoint,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		volAttachments[argIdx] = attachment
+	}
+
+	return volAttachments, nil
+}
+
+// DetachVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) DetachVolumes(args []storage.VolumeAttachmentParams) error {
+
+	for _, attachArg := range args {
+		// Check to see if the volume is already detached.
+		volAttachments, err := s.openstackClient.ListVolumeAttachments(string(attachArg.InstanceId))
+		if err != nil {
+			return err
+		}
+		attachment := findAttachment(attachArg.VolumeId, volAttachments)
+		if attachment == nil {
+			continue
+		}
+
+		err = s.openstackClient.DetachVolume(string(attachArg.InstanceId), attachment.Volume.String())
```

Comment:
Forgive me if I'm being dense, but I don't see an attachment ID being used in the DetachVolume call; only an instance ID and a volume ID.